File: labs/mobile/donor-archive/tunnel/ProxyUriCodec.kt
package com.luminet.android.tunnel

data class CanonicalProxyNode(
    val protocol: String,
    val uuidOrPassword: String,
    val address: String,
    val port: Int,
    val remark: String,
    val params: Map<String, String>
)

class ProxyUriCodec {
    fun parse(rawUri: String): CanonicalProxyNode? {
        try {
            val trimmed = rawUri.trim()
            val schemeIdx = trimmed.indexOf("://")
            if (schemeIdx < 0) return null
            val proto = trimmed.substring(0, schemeIdx).lowercase()
            var rest = trimmed.substring(schemeIdx + 3)

            val remark = if (rest.contains("#")) {
                val p = rest.split("#", limit = 2)
                rest = p[0]
                p[1]
            } else ""

            val (authAndHost, query) = if (rest.contains("?")) {
                val p = rest.split("?", limit = 2)
                Pair(p[0], p[1])
            } else Pair(rest, "")

            val credAndHost = authAndHost.split("@", limit = 2)
            if (credAndHost.size < 2) return null
            val cred = credAndHost[0]
            val hostPort = credAndHost[1].split(":", limit = 2)
            val host = hostPort[0]
            val port = if (hostPort.size > 1) hostPort[1].toInt() else 443

            val params = mutableMapOf<String, String>()
            if (query.isNotEmpty()) {
                query.split("&").forEach { kv ->
                    val pair = kv.split("=", limit = 2)
                    if (pair.isNotEmpty() && pair[0].isNotEmpty()) {
                        params[pair[0]] = if (pair.size > 1) pair[1] else ""
                    }
                }
            }

            return CanonicalProxyNode(proto, cred, host, port, remark, params)
        } catch (e: Exception) {
            return null
        }
    }

    fun serialize(node: CanonicalProxyNode): String {
        val q = node.params.entries.sortedBy { it.key }.joinToString("&") { "${it.key}=${it.value}" }
        val qStr = if (q.isNotEmpty()) "?$q" else ""
        val rStr = if (node.remark.isNotEmpty()) "#${node.remark}" else ""
        return "${node.protocol}://${node.uuidOrPassword}@${node.address}:${node.port}$qStr$rStr"
    }
}

File: labs/control-ui/api-shadow/proxy_uri_codec.ts
/**
 * Unified Proxy URI Codec and Normalizer
 */
export interface CanonicalProxyNode {
  protocol: string;
  uuidOrPassword: string;
  address: string;
  port: number;
  remark: string;
  params: Record<string, string>;
}

export class ProxyUriCodec {
  static parse(rawUri: string): CanonicalProxyNode | null {
    try {
      const trimmed = rawUri.trim();
      const idx = trimmed.indexOf('://');
      if (idx < 0) return null;

      const protocol = trimmed.slice(0, idx).toLowerCase();
      let rest = trimmed.slice(idx + 3);

      let remark = '';
      if (rest.includes('#')) {
        const parts = rest.split('#');
        rest = parts[0]!;
        remark = parts[1] ?? '';
      }

      let query = '';
      if (rest.includes('?')) {
        const parts = rest.split('?');
        rest = parts[0]!;
        query = parts[1] ?? '';
      }

      const atIdx = rest.lastIndexOf('@');
      if (atIdx < 0) return null;
      const cred = rest.slice(0, atIdx);
      const hostPort = rest.slice(atIdx + 1);

      const colonIdx = hostPort.lastIndexOf(':');
      const address = colonIdx >= 0 ? hostPort.slice(0, colonIdx) : hostPort;
      const port = colonIdx >= 0 ? parseInt(hostPort.slice(colonIdx + 1), 10) : 443;

      const params: Record<string, string> = {};
      if (query) {
        for (const kv of query.split('&')) {
          const [k, v] = kv.split('=');
          if (k) params[decodeURIComponent(k)] = v ? decodeURIComponent(v) : '';
        }
      }

      return { protocol, uuidOrPassword: cred, address, port, remark, params };
    } catch {
      return null;
    }
  }

  static serialize(node: CanonicalProxyNode): string {
    const keys = Object.keys(node.params).sort();
    const qParts = keys.map(
      (k) => `${encodeURIComponent(k)}=${encodeURIComponent(node.params[k]!)}`,
    );
    const qStr = qParts.length > 0 ? `?${qParts.join('&')}` : '';
    const rStr = node.remark ? `#${node.remark}` : '';
    return `${node.protocol}://${node.uuidOrPassword}@${node.address}:${node.port}${qStr}${rStr}`;
  }
}
